### PR TITLE
listreceivedbyaddress Filter Address

### PR DESCRIPTION
**Pull request by JeremyRubin**
This gives listreceivedbyaddress the ability to filter for a single address.

This functionality is useful for users such as TumbleBit who need to filter by address.

### CODE REVIEW

File: qa/rpc-tests/receivedby.py [hunk 1]
```diff
@@ -65,10 +65,24 @@ def run_test(self):
...
+        #Only on addr
+        expected = {"address":addr, "account":"", "amount":Decimal("0.1"), "confirmations":10, "txids":[txid,]}
+        res = self.nodes[1].listreceivedbyaddress(0, True, True, addr)
+        assert_array_result(res, {"address":addr}, expected)
+        if len(res) != 1:
+            raise AssertionError("listreceivedbyaddress expected only 1 result")
+
+        #Not on addr
+        other_addr = self.nodes[0].getnewaddress() # note on node[0]! just a random addr
+        res = self.nodes[1].listreceivedbyaddress(5, True, True, other_addr)
```
Comment (jnewbery) [inline qa/rpc-tests/receivedby.py]: Why is minconf 5 here? Better to set it to 0 to match the listreceivedbyaddress call above.

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -1133,6 +1133,9 @@ UniValue ListReceived(const UniValue& params, bool fByAccounts)
     if (params.size() > 0)
         nMinDepth = params[0].get_int();
 
+    bool fFilterAddress = !fByAccounts && params.size() > 3;
+    const CBitcoinAddress filterAddress = fFilterAddress ?  CBitcoinAddress(params[3].get_str()) : CBitcoinAddress{};
+
```
Comment (jnewbery) [inline src/wallet/rpcwallet.cpp]: I think it would be clearer to move these two lines down a bit so the params are being tested in order. I also have a slight preference for an `if params.size() > 3` block to match the existing tests.

File: qa/rpc-tests/receivedby.py (hunk 1 above)
Comment (jnewbery) [inline qa/rpc-tests/receivedby.py]: Why set minconf to 5 here? Can you just set it to 0 to match the call to listreceivedbyaddress() above?

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -1133,6 +1133,9 @@ UniValue ListReceived(const UniValue& params, bool fByAccounts)
     if (params.size() > 0)
         nMinDepth = params[0].get_int();
 
+    bool fFilterAddress = !fByAccounts && params.size() > 3;
+    const CBitcoinAddress filterAddress = fFilterAddress ?  CBitcoinAddress(params[3].get_str()) : CBitcoinAddress{};
```
Comment (jnewbery) [inline src/wallet/rpcwallet.cpp]: nit: I think it's clearer to move these two lines below the `if(params.size() > 2)` block, so the params are being tested in order. I also have a slight preference to change this to a `if(params.size() > 3)` block to match the other tests.

File: src/wallet/rpcwallet.cpp [hunk 4]
```diff
@@ -1178,10 +1185,22 @@ UniValue ListReceived(const UniValue& params, bool fByAccounts)
     // Reply
     UniValue ret(UniValue::VARR);
     map<string, tallyitem> mapAccountTally;
-    BOOST_FOREACH(const PAIRTYPE(CBitcoinAddress, CAddressBookData)& item, pwalletMain->mapAddressBook)
+
+
```
Comment (jnewbery) [inline src/wallet/rpcwallet.cpp]: nit: prefer not to have two blank lines in the middle of a function. Perhaps you can move one of them to between the if block and for block below.

**Comment by jnewbery** [comment]
A couple of general comments:

- How large are the results from the `listreceivedbyaddress` rpc expected to be? If the rpc only ever returns a small number of addresses, it should be easy enough for the client to receive the full list of balances and then filter the list itself?
- if there's a chance that this functionality needs to be extended further to filter on a list of addresses rather than a single address, it'd be better to include that now. Since #8811 , the arguments to the RPCs are part of the API, so changing them later becomes more troublesome. Perhaps change the `only_address` string to a `filter_addresses` array?

I've also added a few nits.

File: qa/rpc-tests/receivedby.py (hunk 1 above)
Comment (JeremyRubin) [inline qa/rpc-tests/receivedby.py]: 👍

File: src/wallet/rpcwallet.cpp (hunk 3 above)
Comment (JeremyRubin) [inline src/wallet/rpcwallet.cpp]: I agree that it's important for the style to be consistent. I opted to convert the other argument parsing in ListReceived to match this style, at the expense of a larger diff, for the benefit of declaring the input parameters const.

**Comment by instagibbs** [comment]
I think (2) isn't really needed since if you need more than a small number you can just do un-filtered or a few repeated calls. I'm no expert on this use-case though.

**Comment by JeremyRubin** [comment]
@EthanHeilman thoughts?


Another optimization would be to allow for caching of this table on construction (maybe keep_cache/clear_cache parameters). This could reduce the O(n*m) complexity for making m repeated calls to O(n + m).

**Comment by NicolasDorier** [comment]
@JeremyRubin The reason why I am interested into that is that here is the code I am using for querying the transactions of a scriptPubKey: [Using listtransactions](https://github.com/NTumbleBit/NTumbleBit/blob/master/NTumbleBit.TumblerServer/Services/RPCServices/RPCBlockExplorerService.cs#L39).

As far as I see this PR would be able to replace my listtransactions nicely. Will review.

File: src/wallet/rpcwallet.cpp [hunk 5]
```diff
@@ -1129,19 +1129,16 @@ struct tallyitem
...
 
-    isminefilter filter = ISMINE_SPENDABLE;
-    if(params.size() > 2)
-        if(params[2].get_bool())
-            filter = filter | ISMINE_WATCH_ONLY;
+    const bool fIncludeWatchOnly = params.size() > 2 && params[2].get_bool();
+    const isminefilter filter =  ISMINE_SPENDABLE | (fIncludeWatchOnly ? ISMINE_WATCH_ONLY : 0);
+
+    const bool fFilterAddress = !fByAccounts && params.size() > 3;
+    const CBitcoinAddress filterAddress = fFilterAddress ?  CBitcoinAddress(params[3].get_str()) : CBitcoinAddress{};
```
Comment (NicolasDorier) [inline src/wallet/rpcwallet.cpp]: `CBitcoinAddress{}` ? shouldn't it be `CBitcoinAddress()` ?

**Comment by EthanHeilman** [comment]
@jnewbery 
>How large are the results from the listreceivedbyaddress rpc expected to be? If the rpc only ever returns a small number of addresses, it should be easy enough for the client to receive the full list of balances and then filter the list itself?

To perform a 800 user mix with TumbleBit we would need a watch list of 1600 addresses. However we there are times in which we only want to learn the status of a single address.

Sorting received transactions by address is a common enough usecase to have an RPC call. It seems likely that people are calling it and then writing filters to select the addresses they want (for instance [this stackexchange question](http://bitcoin.stackexchange.com/questions/14170/is-there-a-json-rpc-method-for-retrieving-all-transactions-for-a-given-address) or [reddit post](https://www.reddit.com/r/Bitcoin/comments/4ww82e/bitcoin_core_rpc_how_to_filter_transactions_by/)). It is an natural addition to the RPC API and one that would make our project and other projects more performant and cleaner.

File: src/wallet/rpcwallet.cpp (hunk 5 above)
Comment (JeremyRubin) [inline src/wallet/rpcwallet.cpp]: should be equivalent; it's c++11 list initializer syntax

**Comment by jnewbery** [comment]
@EthanHeilman - thanks. Sounds like there's widespread demand for this functionality.

concept ACK

File: src/wallet/rpcwallet.cpp (hunk 3 above)
Comment (jnewbery) [inline src/wallet/rpcwallet.cpp]: My personal preference is to keep the if statements rather than converting them to ternary conditionals, as I think that's clearer (and consistent with all the other RPCs). I'm not that concerned about having the parameters const.

File: src/wallet/rpcwallet.cpp [hunk 6]
```diff
@@ -1129,19 +1129,16 @@ struct tallyitem
 UniValue ListReceived(const UniValue& params, bool fByAccounts)
 {
     // Minimum confirmations
-    int nMinDepth = 1;
-    if (params.size() > 0)
-        nMinDepth = params[0].get_int();
+    const int nMinDepth = params.size() == 0 ? 1 : params[0].get_int();
```
Comment (luke-jr) [inline src/wallet/rpcwallet.cpp]: Probably better to leave these alone. If params[0] is null, we really should silently use the default value... The longer version is also more readable/obvious.

File: src/wallet/rpcwallet.cpp [hunk 7]
```diff
@@ -1129,19 +1129,16 @@ struct tallyitem
...
+    const bool fIncludeEmpty = params.size() > 1 && params[1].get_bool();
 
-    isminefilter filter = ISMINE_SPENDABLE;
-    if(params.size() > 2)
-        if(params[2].get_bool())
-            filter = filter | ISMINE_WATCH_ONLY;
+    const bool fIncludeWatchOnly = params.size() > 2 && params[2].get_bool();
+    const isminefilter filter =  ISMINE_SPENDABLE | (fIncludeWatchOnly ? ISMINE_WATCH_ONLY : 0);
+
+    const bool fFilterAddress = !fByAccounts && params.size() > 3;
```
Comment (luke-jr) [inline src/wallet/rpcwallet.cpp]: Should be false if params[3] is null.

File: src/wallet/rpcwallet.cpp [hunk 8]
```diff
@@ -1162,6 +1159,10 @@ UniValue ListReceived(const UniValue& params, bool fByAccounts)
             if (!ExtractDestination(txout.scriptPubKey, address))
                 continue;
 
+            if (fFilterAddress && !(filterAddress.Get() == address)) {
```
Comment (luke-jr) [inline src/wallet/rpcwallet.cpp]: Maybe we should be storing `filterAddress.Get()` above rather than a `CBitcoinAddress`?

File: src/wallet/rpcwallet.cpp [hunk 9]
```diff
@@ -1251,14 +1265,15 @@ UniValue listreceivedbyaddress(const JSONRPCRequest& request)
     if (!EnsureWalletIsAvailable(request.fHelp))
         return NullUniValue;
 
-    if (request.fHelp || request.params.size() > 3)
+    if (request.fHelp || request.params.size() > 4)
         throw runtime_error(
-            "listreceivedbyaddress ( minconf include_empty include_watchonly)\n"
+            "listreceivedbyaddress ( minconf include_empty include_watchonly only_address)\n"
```
Comment (luke-jr) [inline src/wallet/rpcwallet.cpp]: Prefer replacing all the params with an options Object, but perhaps that is better done in a separate PR.

**Review by luke-jr — APPROVED** [review]
utACK, with some nits.

**Comment by JeremyRubin** [comment]
Addressed feedback, and squashed.

@luke-jr it now errors if the passed in address was not a valid address.

File: src/wallet/rpcwallet.cpp (hunk 9 above)
Comment (JeremyRubin) [inline src/wallet/rpcwallet.cpp]: yeah. Separate PR...

File: src/wallet/rpcwallet.cpp [hunk 10]
```diff
@@ -1253,15 +1281,15 @@ UniValue listreceivedbyaddress(const JSONRPCRequest& request)
...
```
Comment (kallewoof) [inline src/wallet/rpcwallet.cpp]: Nit: space before end paren at 'only_address )' to match space after start paren at '( minconf'

Comment (JeremyRubin) [inline src/wallet/rpcwallet.cpp]: > Not sure if I should fix that -- that's how it was before my PR.

**Review by luke-jr — APPROVED** [review]
utACK

File: src/wallet/rpcwallet.cpp [hunk 11]
```diff
@@ -1145,6 +1145,17 @@ UniValue ListReceived(const UniValue& params, bool fByAccounts)
         if(params[2].get_bool())
             filter = filter | ISMINE_WATCH_ONLY;
 
+    bool fFilterAddress = false;
+    CTxDestination filterAddress =  CBitcoinAddress{}.Get();
```
Comment (NicolasDorier) [inline src/wallet/rpcwallet.cpp]: can you use `CNoDestination();` here ? just to be coherent with nulladdress

**Comment by NicolasDorier** [comment]
tested and integrated in NTumbleBit (https://github.com/NTumbleBit/NTumbleBit/commit/cd7c2f42eb7ba1ff071519dd0c2798a1f2fc43d7). This replace `listtransaction *` as I needed.

File: src/wallet/rpcwallet.cpp (hunk 10 above)
Comment (TheBlueMatt) [inline src/wallet/rpcwallet.cpp]: Yea, best to fix, I think.

File: src/wallet/rpcwallet.cpp (hunk 11 above)
Comment (TheBlueMatt) [inline src/wallet/rpcwallet.cpp]: Yes, please.

**Review by TheBlueMatt — COMMENTED** [review]
utACK a96fbed73ed24591dd42845088639da0afaa436a aside from the style nits.

**Comment by JeremyRubin** [comment]
Nits addressed and squashed. Preserved the pre-squash here https://github.com/JeremyRubin/bitcoin/tree/listreceivedbyaddress-filtered-a96fbed73ed24591dd42845088639da0afaa436a

**Comment by NicolasDorier** [comment]
ACK e6f053a66f9196e2d75d15af2f89facd693402ff

File: qa/rpc-tests/receivedby.py [hunk 12]
```diff
@@ -65,10 +65,24 @@ def run_test(self):
...
+        res = self.nodes[1].listreceivedbyaddress(0, True, True, addr)
+        assert_array_result(res, {"address":addr}, expected)
+        if len(res) != 1:
+            raise AssertionError("listreceivedbyaddress expected only 1 result")
+
+        #Not on addr
+        other_addr = self.nodes[0].getnewaddress() # note on node[0]! just a random addr
+        res = self.nodes[1].listreceivedbyaddress(0, True, True, other_addr)
+        if res != []:
+            raise AssertionError("Should not have listed any transactions, got\n%s"%res)
```
Comment (kallewoof) [inline qa/rpc-tests/receivedby.py]: I believe you can use `assert_equal` here too.

File: src/wallet/rpcwallet.cpp [hunk 13]
```diff
@@ -1164,6 +1175,10 @@ UniValue ListReceived(const UniValue& params, bool fByAccounts)
             if (!ExtractDestination(txout.scriptPubKey, address))
                 continue;
 
+            if (fFilterAddress && !(filterAddress == address)) {
```
Comment (kallewoof) [inline src/wallet/rpcwallet.cpp]: Why not `fFilterAddress && filterAddress != address`?

File: src/wallet/rpcwallet.cpp [hunk 14]
```diff
@@ -1253,15 +1281,15 @@ UniValue listreceivedbyaddress(const JSONRPCRequest& request)
     if (!EnsureWalletIsAvailable(request.fHelp))
         return NullUniValue;
 
-    if (request.fHelp || request.params.size() > 3)
+    if (request.fHelp || request.params.size() > 4)
         throw runtime_error(
-            "listreceivedbyaddress ( minconf include_empty include_watchonly)\n"
+            "listreceivedbyaddress (minconf include_empty include_watchonly only_address)\n"
```
Comment (kallewoof) [inline src/wallet/rpcwallet.cpp]: Nit: standard elsewhere is to show optionals as "( opt1 opt2 ... )", not "(opt1 opt2 ...)" (i.e. instead of removing starting space, add ending space)

File: src/wallet/rpcwallet.cpp [hunk 15]
```diff
@@ -2940,7 +2969,7 @@ static const CRPCCommand commands[] =
     { "wallet",             "listaddressgroupings",     &listaddressgroupings,     false,  {} },
     { "wallet",             "listlockunspent",          &listlockunspent,          false,  {} },
     { "wallet",             "listreceivedbyaccount",    &listreceivedbyaccount,    false,  {"minconf","include_empty","include_watchonly"} },
-    { "wallet",             "listreceivedbyaddress",    &listreceivedbyaddress,    false,  {"minconf","include_empty","include_watchonly"} },
+    { "wallet",             "listreceivedbyaddress",    &listreceivedbyaddress,    false,  {"minconf","include_empty","include_watchonly", "only_address"} },
```
Comment (kallewoof) [inline src/wallet/rpcwallet.cpp]: Nit: abide by surrounding style of no space after comma

**Review by kallewoof — COMMENTED** [review]
utACK e6f053a

File: src/wallet/rpcwallet.cpp (hunk 13 above)
Comment (JeremyRubin) [inline src/wallet/rpcwallet.cpp]: (a != b)  is not the same as !(a == b). I don't think != is defined here iirc.

Comment (kallewoof) [inline src/wallet/rpcwallet.cpp]: Not the same only in the sense that the `!=` operator may not be defined, right? Semantically they're identical. And yeah, I see now. `CNoDestination` has no `operator!=`. Ah well.

Comment (JeremyRubin) [inline src/wallet/rpcwallet.cpp]: Except it's not a one liner, it needs to be added for all sorts of classes.

Comment (luke-jr) [inline src/wallet/rpcwallet.cpp]: It's not a `CNoDestination`, it's a `CTxDestination`, which is a boost variant. Some supported versions of boost do not support `operator!=` for variants.

Comment (kallewoof) [inline src/wallet/rpcwallet.cpp]: I thought if all variants accepted the operator the variant would accept it, in which case adding it to `CNoDestination` would be enough, but maybe I'm off on that one.

**Comment by NicolasDorier** [comment]
@JeremyRubin you might need to add a line in client.cpp (https://github.com/bitcoin/bitcoin/issues/9982)

**Comment by NicolasDorier** [comment]
@JeremyRubin Let me know if you are a bit busy and prefer I take care of this PR.

**Comment by JeremyRubin** [comment]
@NicolasDorier Go for it, sorry for the hold up!

**Comment by MarcoFalke** [comment]
Discussion continues at #9991
